### PR TITLE
Paul

### PR DESCRIPTION
**Pull request by Paul44444**
Hallo Lucas,
ich habe einmal so ein Pull request erstellt. Ich hoffe, alles richtig gemacht zu haben. 
Falls ja, kann man anscheinend unter "Files changed" die Änderungen im Vergleich zum "master"-branch einsehen. 
Das sieht natürlich so aus, dass eine Reihe Files ganz gelöscht wurden und dafür eine Reihe neue Files eingefügt wurden.

Viele Grüße!
Paul

### CODE REVIEW

File: README.md
Comment (lubo93) [inline README.md]: You can use quotation marks to highlight file names. Instead of singleNet.py you would have `singleNet.py`.

File: README.md
Comment (lubo93) [inline README.md:18]: Just make it private?

File: README.md
Comment (lubo93) [inline README.md]: I don't understand this sentence, but we can also discuss it tomorrow or in the next days. Would be good to have a clean `README.md`.

File: proj/funcDictionary.py
Comment (lubo93) [inline proj/funcDictionary.py]: Please remove the unused packages (instead of commenting them out).

File: proj/funcDictionary.py
Comment (lubo93) [inline proj/funcDictionary.py]: This info tag is saying nothing and can be removed.

File: proj/funcDictionary.py
Comment (lubo93) [inline proj/funcDictionary.py]: You define `seq` below, so I assume that you don't need that comment anymore?

File: proj/funcDictionary.py
Comment (lubo93) [inline proj/funcDictionary.py]: can be removed?

File: proj/funcDictionary.py
Comment (lubo93) [inline proj/funcDictionary.py:48]: from each string instead of strain?

File: proj/funcDictionary.py
Comment (lubo93) [inline proj/funcDictionary.py:20]: Shouldn't the binning size be a variable and not just simply set to 3?

File: proj/funcDictionary.py
Comment (lubo93) [inline proj/funcDictionary.py:33]: Again a "harcoded" 3 that is confusing. Shouldn't that be a variable instead of a harcoded 3?

File: proj/funcDictionary.py
Comment (lubo93) [inline proj/funcDictionary.py:93]: What is the fourth element? It's not clear why you use the fourth element here.

File: proj/funcDictionary.py
Comment (lubo93) [inline proj/funcDictionary.py]: largest number of nodes instead of "highest" number

File: proj/funcDictionary.py
Comment (lubo93) [inline proj/funcDictionary.py]: `largest_sub_graph`

This whole function should be replaced, because it's not very efficient for determining the largest connected component. Please also don't use the term "subgraph" if you mean connected component. A subgraph can be connected, but it could also be a set of isolated nodes.

Simply use this function `Gc = max(nx.connected_component_subgraphs(G), key=len)` to determine the largest connected component.

See https://networkx.github.io/documentation/networkx-1.10/reference/generated/networkx.algorithms.components.connected.connected_component_subgraphs.html.

File: proj/funcDictionary.py
Comment (lubo93) [inline proj/funcDictionary.py]: largest subgraph

File: proj/funcDictionary.py
Comment (lubo93) [inline proj/funcDictionary.py]: This is super inefficient. Just use something like `sum(degrees)/degrees.size` or  `sum(degrees)/len(degrees)`. Try to avoid `for` loops whenever you can and use built-in functions such as `numpy.sum()`, etc.

File: proj/funcDictionary.py
Comment (lubo93) [inline proj/funcDictionary.py:292]: this is not the degree distribution. it's a list that contains the degree of each node.

File: proj/funcDictionary.py
Comment (lubo93) [inline proj/funcDictionary.py]: please remove if you don't need it and don't use `for` loops to convert lists, etc.

File: proj/funcDictionary.py
Comment (lubo93) [inline proj/funcDictionary.py]: don't use a `for` loop here: `clustersizes = [cluster[i].size for cluster in clusters]` should do the job.

File: proj/funcDictionaryJaccard.py
Comment (lubo93) [inline proj/funcDictionaryJaccard.py:21]: aren't we using the same function already in another file?

File: proj/funcDictionaryJaccard.py
Comment (lubo93) [inline proj/funcDictionaryJaccard.py]: no `for` loops for that.

File: proj/funcDictionaryJaccard.py
Comment (lubo93) [inline proj/funcDictionaryJaccard.py:48]: please use list comprehension and not a loop or use a loop if you want to parallelize later on. it really depends on what you want to do here. you can use `numba` or `pyspark`. 

https://wiki.python.org/moin/ParallelProcessing

**Review by lubo93 — CHANGES_REQUESTED** [review]
Thanks for your work! I would suggest to revise some parts of the code before merging. Some loops are not necessary and I think that we should also already have performance issues in mind. The distance computation with two nested for loops is not efficient. I would suggest numba, pyspark, ... .

You can find some further information here:

https://blog.dominodatalab.com/simple-parallelization/
https://wiki.python.org/moin/ParallelProcessing

File: proj/funcDictionary.py
Comment (Paul44444) [reply proj/funcDictionary.py:20]: "ls" is an array of over all lengths. In a nucleic acid strain, 3 nucleic acids always encode 1 amino acid. Thus the length is always a multiple of 3. That's why the binning size is 3. Otherwise, every value, 
which is not a multiple of 3, would be zero.  This value will always be 3 and is not changed.

File: proj/funcDictionary.py
Comment (Paul44444) [reply proj/funcDictionary.py:33]: As said, it's a biological constant. We can make it a variable, but it is as constant as pi for us.

File: proj/funcDictionary.py
Comment (Paul44444) [reply proj/funcDictionary.py:48]: Yes, with string and strain, I mean the same thing. A strain is just a string of letters. Of course I should decide for
one name, either to say "strain" all the time or to say "string" all the time.

File: proj/funcDictionary.py
Comment (lubo93) [reply proj/funcDictionary.py:33]: Ok, thanks for clarifying. Can we add "Note that dividing by 3 is used to convert nucleic acids to amino acids." somewhere in the documentation? This would be helpful for someone stumbling over this number.

File: proj/funcDictionary.py
Comment (Paul44444) [reply proj/funcDictionary.py:93]: See exlanation in the text

Comment (lubo93) [reply proj/funcDictionary.py:93]: Do you mean strings instead of strains below? The 4th column contains the strings, which are relevant ... ? Then I think that I understand what you mean here.

File: proj/funcDictionary.py
Comment (lubo93) [reply proj/funcDictionary.py:48]: I would suggest string, because "strain" is confusing in a biology setting https://en.wikipedia.org/wiki/Strain_(biology)

**Pull request merged**
